stdlib/public/runtime/EnvironmentVariables.cpp

```diff
@@ -0,0 +1,165 @@
+
+using namespace swift;
+
+// Require all environment variable names to start with SWIFT_
+static constexpr bool hasSwiftPrefix(const char *str) {
+  const char prefix[] = "SWIFT_";
+  for (unsigned i = 0; i < sizeof(prefix) - 1; i++)
+    if (str[i] != prefix[i])
+      return false;
+  return true;
+}
+#define VARIABLE(name, type, defaultValue, help) \
+  static_assert(hasSwiftPrefix(#name), "Names must start with SWIFT");
+#include "EnvironmentVariables.def"
+
+// Define backing variables.
+#define VARIABLE(name, type, defaultValue, help) \
+  type Environment::name ## _variable;
+#include "EnvironmentVariables.def"
+
+// Value parsers. Add new functions named parse_<type> to accommodate more
+// debug variable types.
+static bool parse_bool(const char *name, const char *value, bool defaultValue) {
+  if (!value)
+    return defaultValue;
+  switch (value[0]) {
+  case 'Y':
+  case 'y':
+  case 'T':
+  case 't':
+  case '1':
+    return true;
+  case 'N':
+  case 'n':
+  case 'F':
+  case 'f':
+  case '0':
+    return false;
+  default:
+    swift::warning(RuntimeErrorFlagNone,
+                  "Warning: cannot parse value %s=%s, defaulting to %s.\n",
+                  name, value, defaultValue ? "true" : "false");
+    return defaultValue;
+  }
+}
+
+static uint8_t parse_uint8_t(const char *name,
+                             const char *value,
+                             uint8_t defaultValue) {
+  if (!value)
+    return defaultValue;
+  long n = strtol(value, nullptr, 0);
+  if (n == 0 && errno == EINVAL) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: cannot parse value %s=%s, defaulting to %u.\n",
+                   name, value, defaultValue);
+    return defaultValue;
+  }
+
+  if (n < 0) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: %s=%s out of bounds, clamping to 0.\n",
+                   name, value);
+    return 0;
+  }
+  if (n > UINT8_MAX) {
+    swift::warning(RuntimeErrorFlagNone,
+                   "Warning: %s=%s out of bounds, clamping to %d.\n",
+                   name, value, UINT8_MAX);
+    return UINT8_MAX;
+  }
+
+  return n;
+}
+
+// Print a list of all the environment variables. Lazy initialization makes
+// this a bit odd, but the use of these variables in the metadata system means
+// it's almost certain to run early.
+//
+// The "extra" parameter is printed after the header and before the list of
+// variables.
+static void printHelp(const char *extra) {
+  swift::warning(RuntimeErrorFlagNone, "Swift runtime debugging:\n");
+  if (extra)
+    swift::warning(RuntimeErrorFlagNone, "%s\n", extra);
+#define VARIABLE(name, type, defaultValue, help) \
+  swift::warning(RuntimeErrorFlagNone, "%7s %s [default: %s] - %s\n", \
+                 #type, #name, #defaultValue, help);
+#include "EnvironmentVariables.def"
+}
+
+// Initialization code.
+OnceToken_t Environment::initializeToken;
+
+#if defined(__APPLE__) || defined(__FreeBSD__) || defined(__linux__)
```

Comment:
Windows has `_environ` which is the same variable with a slightly altered spelling.